service/history/workflow/mutable_state_impl_test.go

```diff
@@ -1230,27 +1230,157 @@ func (s *mutableStateSuite) getResetPointsBinaryChecksumsFromMutableState() []st
+				enumsspb.TASK_TYPE_VISIBILITY_DELETE_EXECUTION,
+			},
+		},
+		{
+			name: "upsert close delete",
+			tasks: []tasks.Task{
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.CloseExecutionVisibilityTask{},
+				&tasks.DeleteExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_DELETE_EXECUTION,
+			},
+		},
+		{
+			name: "close delete",
+			tasks: []tasks.Task{
+				&tasks.CloseExecutionVisibilityTask{},
+				&tasks.DeleteExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_DELETE_EXECUTION,
+			},
+		},
+		{
+			name: "delete",
+			tasks: []tasks.Task{
+				&tasks.DeleteExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_DELETE_EXECUTION,
+			},
+		},
+		{
+			name: "start upsert close",
+			tasks: []tasks.Task{
+				&tasks.StartExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.CloseExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_CLOSE_EXECUTION,
+			},
+		},
+		{
+			name: "upsert close",
+			tasks: []tasks.Task{
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.CloseExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_CLOSE_EXECUTION,
+			},
+		},
+		{
+			name: "close",
+			tasks: []tasks.Task{
+				&tasks.CloseExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_CLOSE_EXECUTION,
+			},
+		},
+		{
+			name: "start upsert",
+			tasks: []tasks.Task{
+				&tasks.StartExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_UPSERT_EXECUTION,
+			},
+		},
+		{
+			name: "upsert",
+			tasks: []tasks.Task{
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_UPSERT_EXECUTION,
+			},
+		},
+		{
+			name: "start",
+			tasks: []tasks.Task{
+				&tasks.StartExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_START_EXECUTION,
+			},
+		},
+		{
+			name: "upsert start delete close",
+			tasks: []tasks.Task{
+				&tasks.UpsertExecutionVisibilityTask{},
+				&tasks.StartExecutionVisibilityTask{},
+				&tasks.DeleteExecutionVisibilityTask{},
+				&tasks.CloseExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_DELETE_EXECUTION,
+			},
+		},
+		{
+			name: "close upsert",
+			tasks: []tasks.Task{
+				&tasks.CloseExecutionVisibilityTask{},
+				&tasks.UpsertExecutionVisibilityTask{},
+			},
+			res: []enumsspb.TaskType{
+				enumsspb.TASK_TYPE_VISIBILITY_CLOSE_EXECUTION,
```

Comment:
Is there ever a case where we'll return multiple TaskTypes? If so, could you add tests to ensure our behavior in that case?

Reply:
No, this new logic will always return up to 1 task.

Reply:
Does it still make sense to return a slice then?

Reply:
The function changes the mutable state attribute in-place.